packages/images/src/service.js

```diff
@@ -0,0 +1,199 @@
+  constructor({ port = 4001, protocol = 'http', host = 'localhost', path = './images' }) {
+    this.path = nodePath.resolve(path);
+    this.protocol = protocol;
+    this.port = port;
+    this.host = host;
+    this.database = new DataStore({ path: this.path });
+    this.sourcePath = nodePath.join(this.path, 'source');
+    this.transformsPath = nodePath.join(this.path, 'transforms');
+    fs.ensureDirSync(this.sourcePath);
+    fs.ensureDirSync(this.transformsPath);
+
+    const upload = multer({ dest: this.sourcePath });
+
+    this.app = express();
+
+    this.app.post('/upload', upload.single('image'), async (req, res) => {
+      const { originalname, path } = req.file;
+
+      res.status(201).json(this.storeImageInDatabase({ path, originalname }));
+    });
+
+    this.app.get('/image/:filename', async (req, res) => {
+      const { filename } = req.params;
+      const [id, format] = filename.split('.');
+
+      if (!id || !format) {
+        return res.status(404).send('Not Found (Invalid Filename)');
+      }
+
+      const data = this.database.getImage(id);
+
+      if (!data) {
+        return res.status(404).send('Not Found (No Matching Image)');
+      }
+
+      const originalPath = nodePath.join(this.sourcePath, `${id}.${data.format}`);
+      const resizeOptions = processResizeOptions(req.query);
+
+      if (!resizeOptions && format === data.format) {
+        return res.sendFile(originalPath);
+      }
+
+      let suffix = '';
+
+      if (resizeOptions) {
+        if (resizeOptions.error) {
+          return res.status(400).send(resizeOptions.error);
+        }
+
+        const md5sum = crypto.createHash('md5');
+        md5sum.update(JSON.stringify(resizeOptions));
+        const hash = md5sum.digest('hex');
+
+        suffix = `-${hash}`;
+      }
+
+      let transformedFilename = nodePath.join(this.transformsPath, `${id}${suffix}.${format}`);
+
+      if (!(await fs.exists(transformedFilename))) {
+        try {
+          const sharpImage = sharp(originalPath);
+          if (resizeOptions) {
+            await sharpImage.resize(resizeOptions);
+          }
+          await sharpImage.toFile(transformedFilename);
+        } catch (err) {
+          console.error(err);
+          return res.status(500).send({ error: 'Internal server error' });
+        }
+      }
+
+      res.sendFile(transformedFilename);
+    });
+
+    this.app.get('/image/:id/meta', async (req, res) => {
+      const { id } = req.params;
+      const data = this.database.getImage(id);
+
+      if (!data) {
+        return res.status(404).json({ error: 'not found' });
+      }
+
+      res.json(data);
+    });
+
+    this.app.listen(port, () => {
+      console.log(`🌠 KeystoneJS Image Service ready on port ${port}`);
+    });
+  }
+  getSrc(id, { format, resize = {} }) {
+    let url = `${this.protocol}://${this.host}:${this.port}/image/${id}.${format}`;
+
+    let searchParams = new URLSearchParams();
+    for (let key in resize) {
```

Comment:
I generally try and avoid using `for (.. in ..)` [due to the potential to iterate through inherited values](https://eslint.org/docs/rules/guard-for-in).

Might doing `resize.forEach` be slightly better here?

Reply:
These days I tend not to code so defensively, since this code is going to be called from our own code anyway (see `Implementation.js`) so unless somebody's extending object globals in this process, it's going to be fine. We could go nuts everywhere, or just call that an unsupported use case.

If it's public API, I'd agree more